File: plugins/python/worker.py
import base64
import json
import os
import subprocess
import sys
import time

if endpoint := os.getenv("WEAVEPORT_SOCKET"):
    import socket

    channel = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
    channel.connect(endpoint)
    sys.stdin = channel.makefile("r", encoding="utf-8")
    sys.stdout = channel.makefile("w", encoding="utf-8")

counter = 0


def send(value):
    print(json.dumps(value, separators=(",", ":")), flush=True)


def callback(request, operation, payload):
    callback.counter += 1
    cid = str(callback.counter)
    send(
        dict(
            type="callback",
            id=request["id"],
            callbackId=cid,
            operation=operation,
            payload=payload,
        )
    )
    reply = json.loads(sys.stdin.readline())
    if reply["id"] != request["id"] or reply["callbackId"] != cid:
        raise ValueError("callback identity")
    return reply["value"]


callback.counter = 0


def hang_operation(request):
    payload = request["payload"]
    while True:
        pass


def memory_operation(request):
    payload = request["payload"]
    blocks = []
    while True:
        blocks.append(bytearray(8 * 1024 * 1024))


def callback_flood_operation(request):
    payload = request["payload"]
    for _ in range(payload.get("count", 10)):
        callback(request, "documents.read", {})
    return None


def workspace_operation(request):
    payload = request["payload"]
    path = os.path.join(os.getenv("WEAVEPORT_WORKSPACE", "/tmp"), "state")
    if "text" in payload:
        with open(path, "w") as stream:
            stream.write(payload["text"])
    if not os.path.exists(path):
        return ""
    with open(path) as stream:
        return stream.read()


def execute(request):
    global counter
    payload = request["payload"]
    op = request["operation"]
    if op == "environment":
        return dict(parentSecretPresent="WEAVEPORT_PARENT_CANARY" in os.environ)
    if op == "trace":
        return request["traceId"]
    if op == "echo":
        return payload
    if op == "bulk-map":
        return dict(
            data=base64.b64encode(
                base64.b64decode(payload["data"], validate=True).replace(b"a", b"A")
            ).decode("ascii")
        )
    if op == "search":
        docs = callback(request, "documents.read", payload)
        return sorted(
            [d for d in docs if payload.get("query", "").lower() in d["text"].lower()],
            key=lambda d: d["id"],
        )
    if op == "context":
        return request["context"]
    if op == "counter":
        counter += 1
        return counter
    if op == "reduce":
        return dict(
            state=payload["state"] + payload["amount"],
            events=[dict(kind="changed", at=payload["now"])],
        )
    if op == "delay":
        time.sleep(payload["ms"] / 1000)
        return payload
    if op == "crash":
        os._exit(17)
    if op == "exception":
        raise ValueError("deliberate failure")
    if op == "hang":
        return hang_operation(request)
    if op == "memory":
        return memory_operation(request)
    if op == "oversize":
        return "x" * (2 * 1024 * 1024)
    if op == "malformed":
        print("not-json", flush=True)
        return None
    if op == "callback":
        return callback(request, payload["operation"], payload.get("args", {}))
    if op == "callback-flood":
        return callback_flood_operation(request)
    if op == "workspace":
        return workspace_operation(request)
    if op == "subprocess":
        return subprocess.check_output(["/bin/sh", "-c", "printf child-ok"], text=True)
    raise ValueError("unknown operation")


send(dict(type="ready", protocol=1, pluginVersion=os.getenv("PLUGIN_VERSION", "1")))
for line in sys.stdin:
    request = json.loads(line)
    try:
        send(dict(type="result", id=request["id"], value=execute(request)))
    except Exception:
        send(dict(type="error", id=request["id"], code="plugin-error"))

File: plugins/typescript/worker.ts
import { createConnection } from 'node:net';
import { once } from 'node:events';
import { createInterface } from 'node:readline';
import { existsSync, readFileSync, writeFileSync } from 'node:fs';
import { execFileSync } from 'node:child_process';

type Request = { id: string; operation: string; payload: any; context: any; traceId: string };
const channel = process.env.WEAVEPORT_SOCKET
    ? createConnection(process.env.WEAVEPORT_SOCKET)
    : undefined;
if (channel) await once(channel, 'connect');
const output = channel ?? process.stdout;
const lines = createInterface({ input: channel ?? process.stdin, crlfDelay: Infinity })[
    Symbol.asyncIterator
]();
const workspace = (process.env.WEAVEPORT_WORKSPACE ?? '/tmp') + '/state';
let counter = 0;
let callbackId = 0;
function send(value: unknown): void {
    output.write(JSON.stringify(value) + '\n');
}
async function callback(r: Request, operation: string, payload: unknown): Promise<any> {
    const cid = String(++callbackId);
    send({ type: 'callback', id: r.id, callbackId: cid, operation, payload });
    const reply = JSON.parse((await lines.next()).value!);
    if (reply.id !== r.id || reply.callbackId !== cid) throw new Error('callback identity');
    return reply.value;
}
async function execute(r: Request): Promise<any> {
    const p = r.payload;
    switch (r.operation) {
        case 'environment':
            return { parentSecretPresent: process.env.WEAVEPORT_PARENT_CANARY !== undefined };
        case 'trace':
            return r.traceId;
        case 'echo':
            return p;
        case 'bulk-map': {
            const data = Buffer.from(p.data, 'base64');
            const lowercaseA = 'a'.charCodeAt(0);
            const uppercaseA = 'A'.charCodeAt(0);
            for (let i = 0; i < data.length; i++) {
                if (data[i] !== lowercaseA) continue;
                data[i] = uppercaseA;
            }
            return { data: data.toString('base64') };
        }
        case 'search':
            return (await callback(r, 'documents.read', p))
                .filter((d: any) => d.text.toLowerCase().includes((p.query ?? '').toLowerCase()))
                .sort((a: any, b: any) => a.id.localeCompare(b.id));
        case 'context':
            return r.context;
        case 'counter':
            return ++counter;
        case 'reduce':
            return { state: p.state + p.amount, events: [{ kind: 'changed', at: p.now }] };
        case 'delay':
            await new Promise((resolve) => setTimeout(resolve, p.ms));
            return p;
        case 'crash':
            process.exit(17);
        case 'exception':
            throw new Error('deliberate failure');
        case 'hang':
            while (true) {}
        case 'memory': {
            const blocks: Buffer[] = [];
            while (true) blocks.push(Buffer.alloc(8 * 1024 * 1024, 42));
        }
        case 'oversize':
            return 'x'.repeat(2 * 1024 * 1024);
        case 'malformed':
            output.write('not-json\n');
            return null;
        case 'callback':
            return await callback(r, p.operation, p.args ?? {});
        case 'callback-flood':
            for (let n = 0; n < (p.count ?? 10); n++) await callback(r, 'documents.read', {});
            return null;
        case 'workspace':
            if (p.text !== undefined) writeFileSync(workspace, p.text);
            return existsSync(workspace) ? readFileSync(workspace, 'utf8') : '';
        case 'subprocess':
            return execFileSync('/bin/sh', ['-c', 'printf child-ok'], { encoding: 'utf8' });
        default:
            throw new Error('unknown operation');
    }
}
send({ type: 'ready', protocol: 1, pluginVersion: process.env.PLUGIN_VERSION ?? '1' });
while (true) {
    const line = await lines.next();
    if (line.done) break;
    const request: Request = JSON.parse(line.value);
    try {
        send({ type: 'result', id: request.id, value: await execute(request) });
    } catch {
        send({ type: 'error', id: request.id, code: 'plugin-error' });
    }
}
